gk/main.c

```diff
@@ -542,6 +612,71 @@ setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
 	return ret;
 }
 
+static struct flow_entry *
+find_flow_entry_candidate(struct gk_instance *instance,
+	uint32_t bidx, uint64_t request_timeout_cycles,
+	enum gk_flow_state state_to_add)
+{
+	int32_t index;
+	uint32_t next = 0;
+	const struct ip_flow *key;
+	struct flow_entry *last_fe = NULL;
+	void *data;
+	uint64_t now = rte_rdtsc();
+
+	index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
+		(void *)&key, &data, &bidx, &next);
+	while (index >= 0) {
+		struct flow_entry *fe = &instance->ip_flow_entry_table[index];
+
+		/* Expired flow entry. */
+		if (is_flow_expired(fe, request_timeout_cycles))
+			return fe;
+
+		/*
+		 * Only flow entries with state GK_REQUEST
+		 * will be possibly repaced, others have a higher priority.
+		 */
+		if (fe->state == GK_REQUEST) {
+			uint8_t priority = priority_from_delta_time(now,
+				fe->u.request.last_packet_seen_at);
+			/*
+			 * Do not favor request entries that are not doubling
+			 * its priority when a Gatekeeper server is overloaded.
+			 * We use +2 instead of +1 in the test below to account
+			 * for random delays in the network.
+			 */
+			if (priority > fe->u.request.last_priority + 2)
+				return fe;
+
+			if (state_to_add != GK_REQUEST && (last_fe == NULL ||
+					last_fe->u.request.last_packet_seen_at >
+					fe->u.request.last_packet_seen_at))
+				last_fe = fe;
+		}
+
+		index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
+			(void *)&key, &data, &bidx, &next);
+	}
+
+	return state_to_add == GK_REQUEST ? NULL : last_fe;
```

Comment:
Notice that `last_fe` is initialized with `NULL` and it is only updated if `state_to_add != GK_REQUEST`. So when `state_to_add == GK_REQUEST`, `last_fe` is `NULL`. Therefore, we can simplify the line above to `return last_fe;`.